File: archcive#00-001-100/002-Add Two Numbers/leetcode002.cpp
// <!-- encoding UTF-8 --!>
/*****************************************************************************
 *                     ----Stay Hungry Stay Foolish----                       *
 *   @author :   Shen                                                         *
 *   @name   :   Leetcode 002                                                 *
 *****************************************************************************/

#include <bits/stdc++.h>
using namespace std;
typedef long long int64;
template <class T> inline bool updateMin(T &a, T b) { return a > b ? a = b, 1 : 0; }
template <class T> inline bool updateMax(T &a, T b) { return a < b ? a = b, 1 : 0; }

// Definition for singly-linked list.
// struct ListNode
// {
//     int val;
//     ListNode *next;
//     ListNode(int x) : val(x), next(NULL) {}
// };

class Solution
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        ListNode *head = NULL, *prev = NULL;
        int d1 = 0, d0 = 0;
        while (l1 || l2)
        {
            int hoge = l1 ? l1->val : 0;
            int piyo = l2 ? l2->val : 0;
            int fuga = hoge + piyo + d1;
            d1 = fuga / 10;
            d0 = fuga % 10;
            ListNode *cur = new ListNode(d0);
            if (!head) head = cur;
            if (prev) prev->next = cur;
            prev = cur;
            l1 = l1 ? l1->next : NULL;
            l2 = l2 ? l2->next : NULL;
        }
        if (d1 > 0)
        {
            ListNode *last = new ListNode(d1);
            prev->next = last;
        }
        return head;
    }
};

void test() {}

int main()
{
    test();
    return 0;
}

File: Archive#00-001-100/002-Add Two Numbers/leetcode002.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = prev = ListNode(None)
        d0, d1 = 0, 0
        while l1 or l2:
            hoge, piyo = [_.val if _ else 0 for _ in [l1, l2]]
            fuga = hoge + piyo + d1
            d1, d0 = fuga // 10, fuga % 10
            prev.next = ListNode(d0)
            prev = prev.next
            l1, l2 = [_.next if _ else None for _ in [l1, l2]]
        prev.next = ListNode(d1) if d1 > 0 else None
        return dummy.next
